core/services/directrequest/delegate.go

```diff
@@ -101,11 +123,67 @@ func (d listener) HandleLog(lb log.Broadcast) {
 		return
 	}
 
-	// TODO: Logic to handle log will go here
+	logJobSpecID := lb.RawLog().Topics[1]
+	if logJobSpecID.String() == "0x0000000000000000000000000000000000000000000000000000000000000000" ||
+		logJobSpecID != l.onChainJobSpecID {
+		logger.Debugw("Skipping Run for Log with wrong Job ID", "logJobSpecID", logJobSpecID, "actualJobID", l.onChainJobSpecID)
+		return
+	}
+
+	log := lb.DecodedLog()
+	if log == nil || reflect.ValueOf(log).IsNil() {
+		logger.Error("HandleLog: ignoring nil value")
+		return
+	}
+
+	switch log := log.(type) {
+	case *oracle_wrapper.OracleOracleRequest:
+		l.handleOracleRequest(log)
+		err = lb.MarkConsumed()
+		if err != nil {
+			logger.Errorf("Error marking log as consumed: %v", err)
+		}
+	case *oracle_wrapper.OracleCancelOracleRequest:
+		l.handleCancelOracleRequest(log.RequestId)
+		err = lb.MarkConsumed()
+		if err != nil {
+			logger.Errorf("Error marking log as consumed: %v", err)
+		}
+
+	default:
+		logger.Warnf("unexpected log type %T", log)
+	}
+}
+
+func oracleRequestToMap(req *oracle_wrapper.OracleOracleRequest) map[string]interface{} {
+	result := make(map[string]interface{})
+	result["specId"] = fmt.Sprintf("0x%x", req.SpecId)
+	result["requester"] = req.Requester.Hex()
+	result["requestId"] = fmt.Sprintf("0x%x", req.RequestId)
+	result["payment"] = fmt.Sprintf("%v", req.Payment)
+	result["callbackAddr"] = req.CallbackAddr.Hex()
+	result["callbackFunctionId"] = fmt.Sprintf("0x%x", req.CallbackFunctionId)
+	result["cancelExpiration"] = fmt.Sprintf("%v", req.CancelExpiration)
+	result["dataVersion"] = fmt.Sprintf("%v", req.DataVersion)
+	result["data"] = fmt.Sprintf("0x%x", req.Data)
+	return result
+}
+
+func (l *listener) handleOracleRequest(req *oracle_wrapper.OracleOracleRequest) {
+	meta := make(map[string]interface{})
+	meta["oracleRequest"] = oracleRequestToMap(req)
+	ctx := context.TODO()
+	_, err := l.pipelineRunner.CreateRun(ctx, l.jobID, meta)
```

Comment:
I think we can just use an in-memory queue for this instead of using the database which will be much more scalable (especially for fast chains like matic). So handleOracleRequest just puts something on a buffered channel for some background routines to work on. The log broadcaster handles the restart scenario. We're trying to avoid using the CreateRun/AwaitRun/ResultsForRun because its really slow to do the queuing through the database

Reply:
I guess for the cancelling we could include a ctx in the channel message and use that for the ExecuteAndInsertRun. The listener would need to keep some state (I guess a map) of the enqueued requests to cancel fns and then handleCancelOracleRequest would look up the request and call cancel on it.

Reply:
Don't Runs immediately need to save state as soon as they start running, thus making any attempt to remove use of the DB here inconsequential?

Reply:
The log broadcaster calls HandleLog serially so removing the DB call here speeds up the processing of all logs

Reply:
I agree. I've pretty much come round to the idea of getting rid of the pipeline database queue entirely. Crash-persistence responsibility has moved to log broadcaster and performance is a much bigger concern than we previously imagined.